pilot/pkg/config/kube/ingress/controller.go

```diff
@@ -72,301 +68,244 @@ var schemas = collection.SchemasFor(
-func (c *controller) shouldProcessIngressUpdate(ing *knetworking.Ingress) bool {
-	// ingress add/update
-	shouldProcess := c.shouldProcessIngress(c.meshWatcher.Mesh(), ing)
-	item := config.NamespacedName(ing)
-	if shouldProcess {
-		// record processed ingress
-		c.mutex.Lock()
-		c.ingresses[item] = ing
-		c.mutex.Unlock()
-		return true
-	}
+	// outputs contains all the output collections for this controller.
+	outputs Outputs
 
-	c.mutex.Lock()
-	_, preProcessed := c.ingresses[item]
-	// previous processed but should not currently, delete it
-	if preProcessed && !shouldProcess {
-		delete(c.ingresses, item)
-	} else {
-		c.ingresses[item] = ing
-	}
-	c.mutex.Unlock()
+	status *status.StatusCollections
+}
 
-	return preProcessed
+type Inputs struct {
+	Ingresses      krt.Collection[*knetworking.Ingress]
+	IngressClasses krt.Collection[*knetworking.IngressClass]
+	Services       krt.Collection[*corev1.Service]
+	Nodes          krt.Collection[*corev1.Node]
+	Pods           krt.Collection[*corev1.Pod]
+	MeshConfig     krt.Collection[meshwatcher.MeshConfigResource]
 }
 
-func (c *controller) onEvent(item types.NamespacedName) error {
-	event := model.EventUpdate
-	ing := c.ingress.Get(item.Name, item.Namespace)
-	if ing == nil {
-		event = model.EventDelete
-		c.mutex.Lock()
-		ing = c.ingresses[item]
-		delete(c.ingresses, item)
-		c.mutex.Unlock()
-		if ing == nil {
-			// It was a delete and we didn't have an existing known ingress, no action
-			return nil
-		}
-	}
+type Outputs struct {
+	VirtualServices krt.Collection[config.Config]
+	Gateways        krt.Collection[config.Config]
+}
 
-	// we should check need process only when event is not delete,
-	// if it is delete event, and previously processed, we need to process too.
-	if event != model.EventDelete {
-		shouldProcess := c.shouldProcessIngressUpdate(ing)
-		if !shouldProcess {
-			return nil
-		}
+func NewController(
+	client kube.Client,
+	meshConfig meshwatcher.WatcherCollection,
+	options kubecontroller.Options,
+	xdsUpdater xdsConfigUpdater,
+) *Controller {
+	stop := make(chan struct{})
+	opts := krt.NewOptionsBuilder(stop, "ingress", options.KrtDebugger)
+
+	c := &Controller{
+		client:     client,
+		stop:       stop,
+		status:     &status.StatusCollections{},
+		xdsUpdater: xdsUpdater,
 	}
 
-	vsmetadata := config.Meta{
-		Name:             item.Name + "-" + "virtualservice",
-		Namespace:        item.Namespace,
-		GroupVersionKind: gvk.VirtualService,
-		// Set this label so that we do not compare configs and just push.
-		Labels: map[string]string{constants.AlwaysPushLabel: "true"},
-	}
-	gatewaymetadata := config.Meta{
-		Name:             item.Name + "-" + "gateway",
-		Namespace:        item.Namespace,
-		GroupVersionKind: gvk.Gateway,
-		// Set this label so that we do not compare configs and just push.
-		Labels: map[string]string{constants.AlwaysPushLabel: "true"},
```

Comment:
previously this controller would emit **wrong** events, it would emit hardcoded `item.Name + "-virtualservice"` and `item.Name + "-gateway"` names even though the created `VirtualServices` and `Gateways` dont really have those names, also we could have multiple `VirtualServices` created for a single `Ingress` but we would only emit a single event.

Also, we would add `Labels: map[string]string{constants.AlwaysPushLabel: "true"}`, which caused `needsPush` to always be true, this should no longer be necessary since we now correctly emit events and only emit them when something has really changed.